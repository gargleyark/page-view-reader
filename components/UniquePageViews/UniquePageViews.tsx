import React from "react";
import useLogs from '../../context/LogsContext/useLogs';
import PageViewData from '../PageViewData';

const UniquePageViews = (): JSX.Element => {
  const { logs } = useLogs();

  const pagesWithUniqueIPs = logs.reduce<Record<string, Array<string>>>((prev, { page, ip }) => {
    if (prev[page] && !prev[page].includes(ip)) {
      return { ...prev, [page]: [...prev[page], ip]}
    }
    if (!prev[page]) {
      return { ...prev, [page]: [ip]}
    }

    return prev;
  }, {});

  const uniqueIPCount = Object.entries(pagesWithUniqueIPs).map(([path, ips]) => ({
    path,
    count: ips.length
  }));

  return (
    <div>
      <h2>Unique page views</h2>
      {
        uniqueIPCount
          .sort((a, b) => b.count - a.count)
          .map(({path, count}) => <PageViewData key={path} path={path} count={count} label="unique views" />)
      }
    </div>
  );
}

export default UniquePageViews;