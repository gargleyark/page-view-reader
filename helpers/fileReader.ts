import { Log } from '../context/LogsContext/LogsContext';

export const formatLogs = (logs: string): Array<Log> => {
    return logs
        .split(/\n/)
        .filter(exists => !!exists)
        .map(log => ({
            page: log.split(' ')[0],
            ip: log.split(' ')[1],
        }))
}

export const getFileContent = (file: Blob) => {
    return new Promise<Array<Log>>((resolve) => {
        var fileReader = new FileReader();  
        fileReader.onloadend = () => resolve(formatLogs(`${fileReader.result}`));
        fileReader.onerror = () => resolve([]);
        fileReader.readAsText(file);
    });
};