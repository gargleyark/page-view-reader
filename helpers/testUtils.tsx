import { render } from "@testing-library/react";
import LogsContext, { Log } from '../context/LogsContext/LogsContext'

const Providers = ({ children }: { children: JSX.Element}) => {
  const logs: Array<Log> = [
    {
      page: '/home',
      ip: '1'
    },
    {
      page: '/home',
      ip: '1'
    },
    {
      page: '/about',
      ip: '1'
    },
    {
      page: '/about',
      ip: '1'
    },
    {
      page: '/about',
      ip: '2'
    },
    {
      page: '/about',
      ip: '2'
    }
  ];

  const setLogs = () => {};

  return (
    <LogsContext.Provider value={{ logs, setLogs }}>
      { children }
    </LogsContext.Provider >
  );
};

const customRender = (ui: JSX.Element, options: Record<string, string> = {}) =>
  render(ui, { wrapper: Providers, ...options });

export * from "@testing-library/react";

export { customRender as render };