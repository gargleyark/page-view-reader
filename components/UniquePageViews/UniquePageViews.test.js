import React from "react";
import { render, screen } from "../../helpers/testUtils";
import UniquePageViews from "./";

describe("Unique Page Views", () => {
    it('should render correctly', () => {
        const { container } = render(<UniquePageViews />);
        expect(container).toMatchSnapshot();
    });
    it('should render text on page', () => {
        render(<UniquePageViews />);
        expect(screen.getByText('Unique page views')).toBeInTheDocument();
    });
});