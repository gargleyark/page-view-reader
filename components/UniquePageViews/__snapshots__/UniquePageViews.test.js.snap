// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unique Page Views should render correctly 1`] = `
<div>
  <div>
    <h2>
      Unique page views
    </h2>
    <div>
      <span>
        /about
      </span>
      :
       
      <span>
        2
         
        unique views
      </span>
    </div>
    <div>
      <span>
        /home
      </span>
      :
       
      <span>
        1
         
        unique views
      </span>
    </div>
  </div>
</div>
`;
